File: Handler/GUIHandler/Loader.cs
///Copyright(c) 2015,HIT All rights reserved.
///Summary:Loader for loading GUI
///Author:Irlovan
///Date:2015-11-12
///Description:
///Modification:

using Irlovan.Canal;
using Irlovan.Lib.Symbol;
using Irlovan.Lib.XML;
using Irlovan.Log;
using System.Collections.Generic;
using System.Threading;
using System.Xml.Linq;

namespace Irlovan.Handlers
{
    internal class Loader : GUI
    {

        #region Structure

        /// <summary>
        /// Handler for realtime Event
        /// </summary>
        internal Loader(IServerSession session, LocalInterface.LocalInterface local, GUISource source)
            : base(session, local, source) {
        }

        #endregion Structure

        #region Field

        internal const string Name = "Loader";

        #endregion Field

        #region Function

        /// <summary>
        /// Handler String
        /// </summary>
        /// <param name="session"></param>
        /// <param name="message"></param>
        /// <returns></returns>
        public override bool Handle(IServerSession session, XElement element) {
            ParseSubcription(element);
            return true;
        }

        /// <summary>
        /// Parse Subcription List
        /// </summary>
        private void ParseSubcription(XElement element) {
            XElement sbc = element.Element(SubcriptionPara);
            if (sbc == null) { return; }
            List<string> pathList = new List<string>();
            foreach (var item in sbc.Elements(PagePara)) {
                string pagePath = ParseSubcriptionPage(item);
                if (string.IsNullOrEmpty(pagePath)) { Global.Info.LogRecorder.Log(LogLevelEnum.Warn, Irlovan.Lib.Properties.Resources.LoadGUIFailed + Symbol.NewLine_Symbol + item.ToString()); continue; }
                pathList.Add(pagePath);
            }
            if (pathList.Count == 0) { return; }
            Source.Update(pathList);
            SendGUISource();
        }

        /// <summary>
        /// Send GUI Source
        /// </summary>
        private void SendGUISource() {
            Dictionary<string, XElement> source = Source.GetSource();
            foreach (var item in source) {
                Session.Send(CreateGUIMessage(GetRelativePath(item.Key), item.Value).ToString());
                Thread.Sleep(SendPageDelay);
            }
        }

        /// <summary>
        /// Create GUI Message
        /// </summary>
        /// <returns></returns>
        private XElement CreateGUIMessage(string path, XElement message) {
            XElement result = new XElement(Name);
            XElement sbc = new XElement(SubcriptionPara);
            XElement page = message;
            page.SetAttributeValue(PathAttr, path);
            sbc.Add(page);
            result.Add(sbc);
            return result;
        }

        /// <summary>
        /// Parse Subcription Page
        /// </summary>
        /// <param name="page"></param>
        private string ParseSubcriptionPage(XElement page) {
            string pagePath;
            if (!XML.InitStringAttr<string>(page, PathAttr, out pagePath)) { return null; }
            return pagePath;
        }

        /// <summary>
        /// Dispose
        /// </summary>
        public override void Dispose() {
            base.Dispose();
        }

        #endregion Function

    }
}

File: Web/Web/Script/Handler/GUIHandler/Loader.js
//Copyright(c) 2015,HIT All rights reserved.
//Des:Loader for Loading GUI
//Author:Irlovan   
//Date:2015-03-24

/**Construction**/
Irlovan.Handler.Loader = function (session) {
    this.Name = "Loader";
    this.SubcriptionTag = "SBC";
    this.PageTag = "Page";
    this.PathAttr = "Path";
    this.Session = session;
    this.SBCCallback = null;
    this.LoadCallback = null;
}
/**Handle message from server**/
Irlovan.Handler.Loader.prototype.Handle = function (xDocument) {
    var element = xDocument.documentElement;
    if (element.nodeName != this.Name) { return; }
    this.HandleSubcription(element);

}

/**Subcribe from server**/
Irlovan.Handler.Loader.prototype.Subcribe = function (guiList, callback) {
    this.SBCCallback = callback;
    this.Session.Send(this.CreateLoaderMessage(guiList));
}

/**Load GUI from server**/
Irlovan.Handler.Loader.prototype.Load = function (page, loadCallback) {
    this.LoadCallback = loadCallback;
    var result = {};
    result[page] = false;
    this.Session.Send(this.CreateLoaderMessage(result));
}

/**Handle subcription from server**/
Irlovan.Handler.Loader.prototype.HandleSubcription = function (element) {
    var subcriptions = element.getElementsByTagName(this.SubcriptionTag);
    if ((!subcriptions) || (subcriptions.length == 0)) { return; }
    var pages = subcriptions[0].getElementsByTagName(this.PageTag);
    if ((!pages) || (pages.length == 0)) { return; }
    for (var i = 0; i < pages.length; i++) {
        this.HandleGUIPage(pages[i]);
    }
}

/**Handle gui pages from server by element**/
Irlovan.Handler.Loader.prototype.HandleGUIPage = function (pageElement) {
    var path = pageElement.getAttribute(this.PathAttr);
    if (!path) { return; }
    if (Irlovan.Global.GUICache[path] != null) {
        Irlovan.Global.GUICache[path.toLowerCase()] = pageElement;
    } else {
        if (this.LoadCallback) { this.LoadCallback(pageElement); }
    }
    if (path == Irlovan.Global.PageTo) {
        Irlovan.Global.LoadBar.Off();
        if (this.SBCCallback) {
            this.SBCCallback();
            this.SBCCallback = null;
        }
    }
}

/**Create Loader Init Info**/
Irlovan.Handler.Loader.prototype.CreateLoaderMessage = function (guiList) {
    var loadDoc = Irlovan.Lib.XML.CreateXDocument(this.Name);
    var sbc = Irlovan.Lib.XML.CreateXElement(this.SubcriptionTag);
    loadDoc.documentElement.appendChild(sbc);
    if (!guiList) { return Irlovan.Lib.XML.ToString(loadDoc); }
    for (var gui in guiList) {
        var page = Irlovan.Lib.XML.CreateXElement(this.PageTag);
        page.setAttribute(this.PathAttr, gui);
        sbc.appendChild(page);
    }
    return Irlovan.Lib.XML.ToString(loadDoc);
}

/**Dispose**/
Irlovan.Handler.Loader.prototype.Dispose = function () {
}